src/book/02-system/01-foundations/07-director.mdx

```diff
@@ -0,0 +1,231 @@
+In the event that the kick providers task had a lower priority than this providers task, then it is only able to change this module to run other non `Leaving` providers to achive its causing.
+
+In order for a leaving provider to be run however, it must have first been in the entering or normal states.
+The system cannot immediately jump to a leaving provider.
+
+Example:
+
+```cpp
+on<Provides<Walk>>().then()
+```
+
+### When
+
+A when condition is used to ensure that a provider will only execute when its conditions are met.
+These conditions are provided using a global enum message along with a statement that describes what condition you want.
+For example if you wanted a kick engine to only run if the robot is statically stable or better.
+
+```cpp
+on<Provides<Kick>, When<Stability, std::greater_equal, Stability::STATIC>>().then(...);
+```
+
+These when conditions are updated by using normal NUClear emits.
+
+### Causing
+
+A causing statement is a promise that by running this provider, the desired state will be eventually reached.
+This is used to allow smooth transitions beween modules.
+For example when the walk engine stopped moving it would be in a "Standing" stability state.
+This means the walk engine could have a `Leaving` reaction that caused this state so other modules know it can use this to be standing before they execute their actions.
+
+```cpp
+on<Leaving<Walk>, Causing<Stability, Stability::STANDING>().then(...);
+```
+
+Other providers will take advantage of this in two ways.
+If the other provider has higher priority than this one, it will force this provider group to switch to a `Provides` or `Leaving` that has a causing that satisfies the when condition it is limited on.
+If the other provider has lower priority than this one, it will only be able to force this provider to change to another `Provides` statement and not a leavign one.
+
+## Tasks
+
+Tasks are the jobs that should be executed using the providers.
+Tasks are implemented using NUClears `emit` along with a custom scope `emit<Task>`.
+
+### Root level tasks
+
+When a task is emitted from a reaction which is not a provider, that task is considered a root level task.
+These root level tasks are all considered siblings within the priority comparision system at the root level of the graph.
+When a root level task is emitted with 0 priority it is removed from the list of tasks.
+
+### Subtasks
+
+A subtask is a task that is emitted by a provider.
+When multiple tasks are emitted by a provider, each of these are sibling tasks in the director graph.
+
+### Priority
+
+Priority in the Director is considered based on the closest common anscestor.
+For root tasks this will be the root element.
+Once the closest common anscestor is determined, the priority of the branch we are on is compared to the priority of the branch of the other task.
+The winner gets to use the resources and potentially evict the other task to wait in the queue for the provider group.
```

Comment:
I feel like I need a picture for this whole thing or more in depth explanation.

Reply:
I feel like there is a simpler way of explaining this. The code for it wasn't too hard.
Basically, find your common ancestor, work out which child you are and what the priority is compared to the other child you're competing with. These priorities only come up if two things want the same resources. E.g. walk and kick both want the legs then only one can have it. The one who came from the highest priority parent gets the legs.

Then optional just means if either is optional the other wins and if both are optional then just do the algorithm as normal

Reply:
How do we specify that one comes before the other?